batcher/aligned-batcher/src/lib.rs

```diff
@@ -544,58 +794,110 @@ impl Batcher {
-                error!(
-                    "Failed to serialize verification data: {:?}, resetting batch state",
-                    e
-                );
-                self.flush_queue_and_clear_nonce_cache().await;
+        let gas_price = match self.get_gas_price().await {
+            Some(price) => price,
+            None => {
+                error!("Failed to get gas price");
                 return None;
             }
         };
 
-        // check if the current batch needs to be splitted into smaller batches
-        if current_batch_size > self.max_batch_size {
-            info!("Batch max size exceded. Splitting current batch...");
-            let mut acc_batch_size = 0;
-            let mut finalized_batch_idx = 0;
-            for (idx, (verification_data, _, _, _)) in batch_state.batch_queue.iter().enumerate() {
-                acc_batch_size += match cbor_serialize(verification_data) {
-                    Ok(serialized) => serialized.len(),
+        // Multiply the gas price by 5 to allow for spike in gas price before submitting
+        let gas_price = match gas_price.checked_mul(U256::from(GAS_CHECK_MULTIPLIER)) {
+            Some(price) => price,
+            None => return None, // gas price was too high
+        };
+
+        // Set the batch posting flag to true
+        *batch_posting = true;
+
+        let mut finalized_batch = vec![];
+        let mut batch_queue_copy = batch_state.batch_queue.clone();
+
+        if !self.try_build_batch(&mut batch_queue_copy, &mut finalized_batch, gas_price) {
+            // We cant post a batch since users are not willing to pay the needed fee, wait for more proofs
+            info!("No working batch found. Waiting for more proofs...");
+            *batch_posting = false;
+            return None;
+        }
+
+        // Set the batch queue to batch queue copy
+        batch_state.batch_queue = batch_queue_copy;
+
+        batch_state.update_user_proofs_in_batch_and_min_fee();
+
+        Some(finalized_batch)
+    }
+
+    /// Tries to build a batch from the current batch queue.
+    /// The function iterates over the batch queue and tries to build a batch that satisfies the gas price
+    /// and the max_fee set by the users.
+    /// If a working batch is found, the function tries to make it as big as possible by adding more proofs,
+    /// until a user is not willing to pay the required fee.
+    /// The extra check is that the batch size does not surpass the maximum batch size.
+    /// Note that the batch queue is sorted descending by the max_fee set by the users.
+    /// We use a copy of the batch queue because we might not find a working batch,
+    /// and we want to keep the original batch queue intact.
+    /// Returns true if a working batch is found, false otherwise.
+    fn try_build_batch(
+        &self,
+        batch_queue_copy: &mut PriorityQueue<BatchQueueEntry, BatchQueueEntryPriority>,
+        finalized_batch: &mut Vec<BatchQueueEntry>,
```

Comment:
If this argument is always passed as an empty vector, it may be better to build it inside the function. Instead of a bool, this function could return an `Option<Vec<BatchQueueEntry>>`. If `finalized_batch_works` is true, return `Some(finalized_batch)`, else `None`.

Reply:
changed on 4a78bf951b6686638729e56332e7958eaf3c0949